### PR TITLE
Update usage of pcre2_substitute() for pcre2-10.21

### PR DESCRIPTION
**Pull request by msteed**
- Set `PCRE2_SUBSTITUTE_OVERFLOW_LENGTH` to get the required buffer length from pcre2 instead of guessing. This simplifies the code that calls `pcre2_substitute()`.
- Set `PCRE2_SUBSTITUTE_EXTENDED` to enable extra goodies in the replacement string. This isn't strictly necessary but it may prove useful.

See #2943.

### CODE REVIEW

**Comment by msteed** [comment]
Looks like the CI build is still using pcre2-10.20:

`g++ [...] -I/usr/local/Cellar/pcre2/10.20/include [...]`

**Comment by floam** [comment]
If the CI is building on OS X it using Homebrew, which doesn't have a 10.21 version bump yet.

**Comment by floam** [comment]
The configure script should check for a system provided pcre2 10.21 for symbols it needs - else use the included one?

**Comment by msteed** [comment]
I agree about adding some intelligence to the configure script. I'll see what I can figure out.

I'm not sure how to handle a shared library disappearing. When fish is built against its own copy of pcre2, it doesn't build a shared library.

**Comment by floam** [comment]
Ideally, in the case where it is using a shared library, it'd be linked in weakly. Instead of blowing up immediately, fish can run - fish `string` would check if some `pcre2_function() == NULL`. If so, print a TERRIBLE error, return form the function, but don't crash.

**Comment by msteed** [comment]
Yes, that might indeed be awesome, but it's outside the scope of this PR.

File: configure.ac [hunk 1]
```diff
@@ -817,8 +818,27 @@ if test "x$included_pcre2" != "xyes"; then
     # and so AC_CHECK_LIB won't work (can't use a variable as library name)
     # AC_SEARCH_LIBS will use the existing $LIBS flags with no additional library first
     AC_SEARCH_LIBS([pcre2_compile_$WCHAR_T_BITS], [],
-      [ working_pcre2=yes
-        AC_MSG_NOTICE([using system PCRE2 library])
+      [ # pcre2 lib found, check for minimum version
+        AC_MSG_CHECKING([for PCRE2 $MIN_PCRE2_VERSION or later])
+        PCRE2_VERSION=`$PCRE2_CONFIG --version`
+        AX_COMPARE_VERSION([$PCRE2_VERSION], [ge], [$MIN_PCRE2_VERSION],
```
Comment (zanchey) [inline configure.ac]: Any reason not to use `AS_VERSION_COMPARE` which is built in? See the Doxygen check.

**Comment by zanchey** [comment]
> If the CI is building on OS X (news to me!)

Indeed it does, with many thanks to Travis CI. It only tests the autotools build at this stage.

**Comment by krader1961** [comment]
One of us should check that our Travis CI is setup to build and test fish on the intersection between platforms we support and Travis supports. I'm just about ready to ask for a review of my IWYU (include-what-you-use) cleanup and knowing that my change builds and runs correctly on more platforms than I have at my disposal would be great.

File: configure.ac (hunk 1 above)
Comment (msteed) [inline configure.ac]: No reason except that I am an autotools newbie :-)

Comment (zanchey) [inline configure.ac]: No worries - let's go with that as it avoids the aclocal mess.

**Comment by zanchey** [comment]
Merged with some edits and rebasing:

To git@github.com:fish-shell/fish-shell.git
   1f06e5f..c2f9d60  master -> master

Many thanks!

**Pull request closed without merging**